File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.java
class Solution {
    void solve(int[] nums, int i, int target, List<Integer> ds, List<List<Integer>> res) {
       if(target == 0) {
            res.add(new ArrayList<>(ds));
            return;
        }
        if(target < 0 || i >= nums.length) {
            return;
        }
        
        for(int ind = i; ind < nums.length; ind++) {
            if(ind > i && nums[ind] == nums[ind-1]) continue;
            ds.add(nums[ind]);
            solve(nums, ind+1, target-nums[ind], ds, res);
            ds.remove(ds.size()-1);
        }
    }
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> ds = new ArrayList<>();
        Arrays.sort(candidates);
        solve(candidates, 0, target, ds, res);
        return new ArrayList<>(res);
    }
}

File: Data Structures & Algorithms/permutations/submission-1.cpp
class Solution {
public:
   void solve(vector<int>& nums, vector<vector<int>>& res, int n, int idx) {

        if(idx >= n) {
            res.push_back(nums);
            return ;
        }

        for(int i = idx; i < n; i++) {
            swap(nums[i], nums[idx]);
            solve(nums, res, n, idx+1);
            swap(nums[i], nums[idx]);
        }

   }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res; 
        int n = nums.size();
        solve(nums, res, n, 0);
        return res;
    }
};
